sale_promotion_discount_in_field/models/sale_order.py

```diff
@@ -0,0 +1,161 @@
+# Copyright 2022 Ooops404
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    def _get_applicable_no_code_promo_program(self):
+        """
+        Override method for return programs other than discount line reward type
+        """
+        self.ensure_one()
+        return (
+            super()
+            ._get_applicable_no_code_promo_program()
+            .filtered(
+                lambda program: program.reward_type != "discount_line",
+            )
+        )
+
+    def _get_applicable_no_code_promo_discount_line_program(self):
+        """
+        Return actual programs with reward type is discount line
+        """
+        self.ensure_one()
+        programs = (
+            self.env["coupon.program"]
+            .with_context(
+                no_outdated_coupons=True,
+                applicable_coupon=True,
+            )
+            .search(
+                [
+                    "&",
+                    ("reward_type", "=", "discount_line"),
+                    ("promo_code_usage", "=", "no_code_needed"),
+                    "|",
+                    ("rule_date_from", "=", False),
+                    ("rule_date_from", "<=", self.date_order),
+                    "|",
+                    ("rule_date_to", "=", False),
+                    ("rule_date_to", ">=", self.date_order),
+                    "|",
+                    ("company_id", "=", self.company_id.id),
+                    ("company_id", "=", False),
+                ]
+            )
+            ._filter_programs_from_common_rules(self)
+        )
+        return programs
+
+    def _create_new_no_code_promo_reward_lines(self):
+        """
+        Override core method to save program with new reward rule
+        """
+        self.ensure_one()
+        super()._create_new_no_code_promo_reward_lines()
+        order = self
+        programs = order._get_applicable_no_code_promo_discount_line_program()
+        programs = (
+            programs._keep_only_most_interesting_auto_applied_global_discount_program()
```

Comment:
Should we add this type of promotion to the `_is_global_discount_program` rules?

Reply:
I think that's not necessary

Reply:
But then two discount promotions could be applied altogether. And even if only one has effects on the lines, the others would be linked to the sales order anyway.

This is an example where I defined two discount in fields promos. The two are linked, but only the last one is applied:
![image](https://user-images.githubusercontent.com/5040182/176840841-d633c313-2476-4dcb-b573-2f8e57ee85ed.png)
![image](https://user-images.githubusercontent.com/5040182/176840857-03485fc1-2d41-40e4-9f27-b352ecf208d7.png)

It would be something like:

```python
    def _is_global_discount_program(self):
        discount_in_field_global = (
            self.promo_applicability == "on_current_order"
            and self.reward_type == "discount_line"
            and self.discount_apply_on == "on_order"
        )
        return super()._is_global_discount_program() or discount_in_field_global
```

Reply:
@GabbasovDinar what do you think ?

Reply:
In your case, after overriding the _is_global_discount_program method, I can remove the unnecessary code and it will work the same way. Now with your example I have only one applied discount program

Thank you, please check :)